File: problems/0001-two-sum/python/two_sum.py
"""LeetCode 1. Two Sum."""


def two_sum(nums: list[int], target: int) -> list[int]:
    """Indices of the two numbers that add up to ``target``.

    Single pass with a ``value -> index`` dict. O(n) time, O(n) space.
    """
    seen: dict[int, int] = {}
    for i, n in enumerate(nums):
        j = seen.get(target - n)
        if j is not None:
            return [j, i]
        seen[n] = i
    return []

File: problems/0001-two-sum/kotlin/src/main/kotlin/TwoSum.kt
// LeetCode 1. Two Sum
//
// Indices of the two numbers that add up to `target`.
// Single pass with a `value -> index` map. O(n) time, O(n) space.
fun twoSum(nums: IntArray, target: Int): IntArray {
    val seen = HashMap<Int, Int>()
    for ((i, n) in nums.withIndex()) {
        seen[target - n]?.let { return intArrayOf(it, i) }
        seen[n] = i
    }
    return intArrayOf()
}

fun main() {
    println(twoSum(intArrayOf(2, 7, 11, 15), 9).toList())
}
